Adding IEEE802.11ah compatibility

irongiant33 opened the pull request:
Hi Bastian,

I have been working on building a SDR receiver for Wi-Fi HaLow, a.k.a. IEEE 802.11ah. You can find my documentation and some GRC flowgraphs here: [https://github.com/irongiant33/gr-halow](https://github.com/irongiant33/gr-halow). 

The implementation is far from complete, but I have been using gr-ieee802-11 as the basis for my changes. By opening this pull request, I hope you'll see something that I am missing in terms of getting a fully running receive flow. My current approach breaks your implementation of IEEE 802.11a/g, but I intend to go back and ensure gr-ieee802-11 is fully compatible with 802.11a/g/ah when my code actually works. I just thought it would be easiest to modify in place to compare the diff first before making a separate PR for the addition of 802.11ah functionality.

## Why?

I think that it is way more achievable for a real-time SDR implementation of 802.11ah than 802.11a/g, especially with more affortable SDRs like the PlutoSDR or HackRF. Unlike the high bandwidth and data rate for 802.11a/g, 802.11ah operates at 10x lower bandwidth/rates than 802.11a. Instead of 20, 40, 80 and 160 MHz bandwidths, HaLow operates with 1, 2, 4, 8, and 16 MHz channels. The devices I have currently only support the 1, 2, and 4 MHz channels.

If the purpose of HaLow is extending mesh network reach as well as number of devices available on the network, I think adding SDR transceiver functionality can expose more hobbyists and developers to its benefits and make HaLow more accessible. 

## Current Approach

I have been reading upon IEEE Std 802.11-2020 to compare your implementations of 802.11a/g with 802.11ah. So far, you can see I mostly have been redefining constants except for the frame equalizer implementation where I am actually trying to pull out information from the SIG field of the PPDU. For the time being, I am considering that even pulling data from the SIG field is a win. At that point, I would ensure compatibility with 802.11a/g before making a pull request here and then going on to add MAC logic to the receive chain for getting information from the data field.

## Lingering Questions/Problems

I'll add comments directly to code as necessary, but in general, if you see anywhere I might be falling short based on your implementation of ieee802-11 I'd appreciate your insight

irongiant33 committed:
experimenting with small change to make sure no issue with pybind

irongiant33 committed:
updating frame equalizer with constants that are relevant to halow. Parsing signal field differently.

irongiant33 committed:
adjusting length of the LONG field to match number of samples per OFDM signal in halow

irongiant33 committed:
updating length of inputs and outputs of the frame equalizer based off of what halow expects

irongiant33 committed:
updating constants in sync long and sync short files

irongiant33 commented:
Where do you get this array from?  https://github.com/bastibl/gr-ieee802-11/pull/412/files#diff-91c264c8b16fa2e3de166f93e05cd89098c6e02e45f90e8f1344848ba73629d6L257-L290

 Initially, I thought it was from p.4170 of the 802.11-2020 spec, but that section only corresponded to this array: https://github.com/bastibl/gr-ieee802-11/pull/412/files#diff-4dc901ae703564f5a6774a2f7c6fcdc47eb988e5182bd3a9e6957667b46ee7c9L24-L28

samuelmiller97 committed:
modifying some more constants

bastibl commented:
Hi, great to see work on HaLow! The approach makes sense. Unfortunately, the link to the array in question does not seem to work for me. Maybe you could link to the current file and not the PR, which might change if you push more commits.

irongiant33 committed:
adding comment above unknown array

irongiant33 commented on lib/sync_long.cc:
```diff
@@ -254,6 +254,7 @@ sync_long::sptr sync_long::make(unsigned int sync_length, bool log, bool debug)
     return gnuradio::get_initial_sptr(new sync_long_impl(sync_length, log, debug));
 }
 
+//where does this array come from?
```
sorry that the link disappeared, the array below is the one I am not sure where it comes from in the standard.

irongiant33 replied:
essentially, I am trying to get down to the root of the HaLow synchronization problem that I am trying to discuss here: [https://github.com/irongiant33/gr-halow/blob/main/docs.md#wifi-synchronization](https://github.com/irongiant33/gr-halow/blob/main/docs.md#wifi-synchronization). For some reason, my modifications to sync long and sync short are not producing output and I'm trying to figure out why. No point in trying to fix the frame equalizer yet if no data ever reaches it haha

bastibl replied:
See this script: https://github.com/bastibl/gr-ieee802-11/blob/maint-3.10/utils/sync_words.py IIRC, in the standard the preambles are only specified in frequency domain.

irongiant33 replied:
you are correct, but the file seems to be this R script instead of the Python script you mentioned. [https://github.com/bastibl/gr-ieee802-11/blob/maint-3.10/utils/create_long.R](https://github.com/bastibl/gr-ieee802-11/blob/maint-3.10/utils/create_long.R). I'll keep digging from here and let you know if I have some more questions, thanks!

irongiant33 committed:
modifying include to enable this to buid

irongiant33 committed:
modifying synchronization sequences for halow

irongiant33 committed:
updating decode MAC with HaLow constants

irongiant33 committed:
troubleshooting signal field decoding

irongiant33 commented on lib/sync_long.cc:
```diff
@@ -47,7 +47,8 @@ class sync_long_impl : public sync_long
+                                   //also 320 samples for ieee802.11ah (160us for short field, 160us for long field @1MHz)
     {
 
         set_tag_propagation_policy(block::TPP_DONT);
```
@bastibl  What was your intention for the sync_short and sync_long blocks? Was it for sync_short to perform coarse frequency correction and roughly labels the start of the wifi frame on the short training sequence, and then sync_long does fine frequency synchronization and timing correction then re-tags the start of the wifi frame to be more accurate on the long training sequence? Hence the tag propagation policy of "do not propagate".

Reason for asking is because I believe that I may be getting thrown off by the IEEE specification overloading the terms "SHORT" and "LONG". I reviewed the specification for clauses 17 (802.11a) and 19 (802.11g) and I think that short and long in your context (802.11a/g) refer to the training sequences in the preamble, but 802.11ah it has multiple meanings.

Where the "short" and "long" terms get overloaded with HaLow is on p.3197, section 23.3.2 of the specification "S1G PPDU format" that defines three formats for the S1G PHY: S1G_SHORT, S1G_LONG, and S1G_1M. Each of these formats then has a short and long training sequence.

On the bright side, it is looking like I am roughly getting close to synchronization after sync_short, but I might need to dig into sync_long some more to figure out why it looks so weird.
- [Time plots after the sync_short and sync_long blocks showing the correct number of tags, as well as correct tag placement](https://github.com/irongiant33/gr-halow/commit/dc4fa776d400534c800b1a0154d3207f93ba137d#diff-cd7f889aec9daee7d8b66f9d9a20808a0f6973a423a8c1b6538b233765853837)
- [Zoomed in time plot showing that sync_short appears to be roughly synchronized, whereas sync_long looks to be off](https://github.com/irongiant33/gr-halow/commit/4778a0676881dbc1a4d7aa0d23b4f66284bb785f#diff-f47bf341a7c8f864f5ccc3d0999f6decda01b4bb7f9f4168da8e7747db900ee7)

bastibl replied:
Short is for frame detection and a first frequency offset correction. It uses the plateau of the auto correlation of the short training field to trigger frame detection. It's computationally efficient but not precise.

Once a frame is detected, Long searches for the precise frame start by cross correlating with the long training symbol and aligns the FFT window accordingly. It also removes the cyclic prefix, does another frequency offset correction.

irongiant33 committed:
continuing to hunt down the reason why the signal field is not producing consistent values

Commit:
sync short : modified constants

Commit:
viterbi decoder : added bpsk 1/2 x2 (mcs 10)

Commit:
mapper : added bpsk 1/2 x2 (mcs 10)

Commit:
equalizer : resolved wrong definition of LTS in frequency domain

Commit:
equalizer : added function get_csi_at

Commit:
equalizer LS : compute CSI based on an average of all four LTS

Commit:
utils : define/modify utils functions, create new constants

Commit:
sync long : resolved bugs in frequency offset computation

Commit:
sync long : fixed GI filtering (temporary fix) and added comments

Commit:
frame equalizer : corrected pilots definition

Commit:
frame equalizer : rewrite carrier frequency offset and sampling offset compensation iaw corrected pilots definition

Commit:
frame equalizer : decode (deinterleave, unrepeat, BCC decode) and parse (split/interpret SIG bits and compute CRC) SIG field

Commit:
adapted tags passed to the mac

Commit:
frame equalizer : forward complex samples downstream to allow for maximum ratio combining of mcs 10 frames

Commit:
moved functions for decoding to utils so it can be shared to frame_equalizer and decode_mac

Commit:
decode_mac - intermediary commit

Commit:
adapted mapper.h python binding

Commit:
added support for traveling pilots

Commit:
fix error in copy paste from last commit

Commit:
fix bug in sample passing from frame_equalizer to decode_mac

Commit:
decode_mac : correct service field length

Commit:
adding support at PHY layer for most halow mcs (part 1) + adding decoding at MAC layer of S1G beacon

Commit:
frame_equalizer : fixed sampling offset computation

Commit:
removed some debugging code

Commit:
adding support at PHY layer for most halow mcs (part 2)

Commit:
mapper : implemented support for halow frame mapping

Commit:
signal_field : implemented support for halow sig field

Commit:
equalizer : added snr computation

Commit:
equalizer : corrected get_csi function

Commit:
added some comments

Commit:
added scripts to generate pilots and sync words

Commit:
modified tx, rx, loopback and phy_hier graphs

Commit:
viterbit_decoder_generic : deleting BPSK_3_4, adding BPSK_1_2_REP

Commit:
removing comments

Commit:
frame_equalizer : do not produce any output if in case of bad sig field crc

Commit:
corrected a bug where as from the 127th symbol, the pilot generated had inverted polarity

Commit:
mapper : added a missing free()

Commit:
decode_mac : resolved bug in print_output() function

Commit:
remove some debugging comments

Commit:
changing min plateau

Commit:
replace block's name WiFi with HaLow

Commit:
removed unnecessary comments

Commit:
added frame symbols, bytes and encoding computation

Commit:
removed old files

Commit:
removing unnecessary comments

Commit:
adding credits file

Commit:
Revert "adding credits file"

This reverts commit d8d5500a2c72229e5216d3e6c0173d665b29696b.

Commit:
added credits

irongiant33 committed:
Merge pull request #1 from neo-knight-td/maint-3.10-halow-1M

802.11ah WiFi Halow 1M - Full Transceiver

samuelmiller97 committed:
renaming example files to avoid build errors

Commit:
add rx verification script

irongiant33 committed:
further validating implementation of 802.11ah, adding script to automatically build and install

irongiant33 committed:
Merge pull request #2 from irongiant33/neo-knight-td_pull-request

neo-knigh-td

irongiant33 committed:
making small change to README so PR can be initiated

neo-knight-td committed:
added script to clean and uninstall

neo-knight-td committed:
Modified README with (un)/installation and usage guidelines

neo-knight-td committed:
adapted branch name in README

irongiant33 committed:
Merge pull request #3 from irongiant33/backwards-compatibility

making small change to README so PR can be initiated

irongiant33 commented:
Hi @bastibl 

Thanks to the help of @neo-knight-td , we have implemented a 802.11ah full 1MHz receiver based on your original code. We are ready to merge our changes into your repository, but due to the differences between 802.11ah and 802.11a/g/p, we thought the code would best live in a `maint-3.10-802-11-ah` branch instead of the `maint-3.10` branch. Would you be able to create this new branch and redirect this PR to merge there instead before I unmark this as draft?

For convenience, we have also added some easy scripts to enable cleaning the build and artifacts for easy switching between gr-ieee80211 compatibility with a/g/p and with ah - all documented here: https://github.com/irongiant33/gr-ieee802-11?tab=readme-ov-file#installation-of-the-maint-310-802-11-ah-branch. 

What are your thoughts? Let us know if you have any troubles running the examples.  If you're not interested in running the code, you can follow along with our verification of the decoding ability here: https://github.com/irongiant33/gr-ieee802-11/pull/2

samuelmiller97 committed:
Merge branch 'maint-3.10' of https://github.com/bastibl/gr-ieee802-11 into maint-3.10

samuelmiller97 committed:
reverting back to Bastian's current working changes

samuelmiller97 committed:
adding reference to gr-ieee802-11ah in the README

bastibl commented:
I cherry-picked the change to the readme. Thanks
